misc: Split up tests in daily-tests.yaml

mkjost0 opened the pull request:
This splits up the gem5 library example tests by Suite UID, as right now running them together uses the runner for a long period of time.  It is important to note that doing this means additional tests from this directory will need to be manually added, such as the kvm tests.

Change-Id: Ib2a0aca08f9b51b60e9dd0528324372cf2d98c05

mkjost0 committed:
misc: Split up tests in daily-tests.yaml

This splits up the gem5 library example tests by Suite UID, as
right now running them together uses the runner for a long
period of time.  It is important to note that doing this means
additional tests from this directory will need to be
manually added, such as the kvm tests.

Change-Id: Ib2a0aca08f9b51b60e9dd0528324372cf2d98c05

powerjg commented on .github/workflows/daily-tests.yaml:
```diff
@@ -363,9 +365,189 @@ jobs:
+      if: success() || failure()
+      uses: actions/upload-artifact@v3
+      env:
+        MY_STEP_VAR: ${{github.job}}_COMMIT.${{github.sha}}_RUN.${{github.run_id}}_ATTEMPT.${{github.run_attempt}}
```
Does this say "gem5_library_example_tests_x86_ubuntu_run_ALL_x86_64_opt" somewhere in the filename?

mkjost0 replied:
Yeah, ${{github.job}} does that, so we don't have duplicate names in the artifact output.

powerjg commented:
Let's clean up this file by using matrix before merging.

mkjost0 committed:
misc: Refactor daily-tests.yaml

This changes the daily tests to use a matrix in order to run
tests.  It also includes forces the cleaning step to run
regardless of success or failure.  With this refactoring, now
all builds of gem5 must finish before any tests run, and all
tests download all artifacts from all the build runs.

Change-Id: I16e1bc9acaf619feb85fba53eb6129e7df3fe409

mkjost0 committed:
misc: Move gem5 library example tests into a matrix

This moves the gem5 library example tests into a separate matrix,
so they can run on separate runners

Change-Id: Ie9f51b5bae9e7e424d1c98b545b4cf92b481a2fb

BobbyRBruce commented on .github/workflows/daily-tests.yaml:
```diff
@@ -172,6 +172,57 @@ jobs:
+    - run: chmod u+x build/ALL_MESI_Two_Level/gem5.opt
+    - name: long ${{ matrix.test-type }} gem5_library_example_tests
+      working-directory: ${{ github.workspace }}/tests
+      run: ./main.py run --uid SuiteUID:tests/gem5/gem5_library_example_tests/test_gem5_library_examples.py:test-gem5-library-example-${{ matrix.test-type }} --length=long --skip-build -vv
```
I don't like how brittle this is, as if we rename any of these directories or even the test, things will break, but I can't see a better way to do it, so I won't block on this. It's just something to keep in mind.

BobbyRBruce commented on .github/workflows/daily-tests.yaml:
```diff
@@ -8,404 +8,148 @@ on:
       - run: echo "This job's status is ${{ job.status }}."
 
+  # NULL build added separately since NULL is a keyword in yaml files
   build-gem5-NULL:
```
You explained to be earlier that you separated this out to it's own separate build stop because "NULL" is keyword. Are you sure you can't just put this as a string in the `image` matrix? E.g., `image: ["NULL", "ALL"...`.

powerjg commented on .github/workflows/daily-tests.yaml:
```diff
@@ -8,404 +8,128 @@ on:
+  build-gem5:
+    strategy:
+      matrix:
+        image: [ALL, ALL_CHI, ARM, ALL_MSI, ALL_MESI_Two_Level, "NULL", NULL_MI_example, RISCV, VEGA_X86]
```
Can you add a comment on why `"NULL"` needs quotes?

powerjg commented on .github/workflows/daily-tests.yaml:
```diff
@@ -8,404 +8,128 @@ on:
+    strategy:
+      matrix:
+        image: [ALL, ALL_CHI, ARM, ALL_MSI, ALL_MESI_Two_Level, "NULL", NULL_MI_example, RISCV, VEGA_X86]
+        include:
```
A comment here would be helpful. I think I get it, but someone new to this file may be confused.

powerjg commented on .github/workflows/daily-tests.yaml:
```diff
@@ -8,404 +8,128 @@ on:
 
-
-  # This runs the unit tests for the build/ALL/unittests.debug build.
   unittests-all-debug:
```
This is not both debug and fast

powerjg commented on .github/workflows/daily-tests.yaml:
```diff
@@ -8,404 +8,128 @@ on:
     container: gcr.io/gem5-test/ubuntu-22.04_all-dependencies:latest
-    needs: [build-gem5, build-gem5-ALL_CHI]
+    needs: [name-artifacts, build-gem5]
     timeout-minutes: 1440 # 24 hours
```
Is this 24 hours *each* or total?

powerjg commented on .github/workflows/daily-tests.yaml:
```diff
@@ -8,404 +8,128 @@ on:
         # Scheduled workflows run on the default branch by default. We
         # therefore need to explicitly checkout the develop branch.
         ref: develop
+    # download all artifacts for each test
```
Do you need to download all of them for all tests? If so, can you explain why in a comment?

mkjost0 replied:
This one I'll explain here as well, but because we're now using a matrix, none of the long tests can start until all of the versions of gem5 within the matrix have been built.  I was toying around with the idea of still only downloading the necessary artifacts for each test, but the ways I was thinking of to do that would make this file a lot less readable than if I were to just download everything, plus it doesn't take significantly longer to wait for everything else to download.  There were also marketplace actions to download multiple artifacts in one step (which I think using would allow me to only download necessary artifacts but in a readable way), but they weren't from GitHub so I opted not to use those either

powerjg reviewed (commented):
Generally, this looks good. It's a bit confusing, so I think it could use more comments before merging.

BobbyRBruce commented:
I think you've accidently included [cpu-kvm: Make using perf when using KVM CPU optional](https://github.com/gem5/gem5/pull/105/commits/188d3390d7312a1ae399a7661c9262f8958de29f) into this PR. I'd do `git rebase -i HEAD~5` and remove this change then force push with `git push -f`.

BobbyRBruce reviewed (approved):
I think this is fine. I don't like the 'name-artifacts' job but I accept this may be the easiest way to do it.
